src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -884,7 +884,24 @@ private static IEnumerable<TSource> TakeWhileIterator<TSource>(IEnumerable<TSour
         public static IEnumerable<TSource> Skip<TSource>(this IEnumerable<TSource> source, int count)
         {
             if (source == null) throw Error.ArgumentNull("source");
-            return SkipIterator<TSource>(source, count);
+
+            IList<TSource> sourceList = source as IList<TSource>;
+            return sourceList != null ? SkipList(sourceList, count) : SkipIterator<TSource>(source, count);
+        }
+
+        private static IEnumerable<TSource> SkipList<TSource>(IList<TSource> source, int count)
+        {
+            if (count < 0)
+            {
+                count = 0;
+            }
+
+            int sourceCount = source.Count;
```

Comment:
> @stephentoub, any objection?

No objection, thanks.  My suggestion was based on performance, but that loses out to correctness.  The only way it would be incorrect is in a situation where there's already invalid usage, but I'm fine with the argument that we want to be as correct as possible even in that situation, given some definition of "correct" (there will still be oddities, and things that may be considered incorrect).